# 01. Importing libraries

In [76]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# 02 Importing data

In [79]:
# Establish path
path = r'C:\Users\elena\Documents\02.2025 Instacart Basket Analysis'

In [81]:
# Import Products datafile
df_prods = pd.read_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'), index_col = False)

In [83]:
# Check products dataframe 
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [85]:
# Import orders_products_combined datafile 
df_opcombined = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined2.pkl'))

In [87]:
# Check the previously merged dataframe
df_opcombined.head()

order_id  product_id  add_to_cart_order  reordered  user_id  order_number  \
0         2       33120                  1          1   202279             3   
1         2       28985                  2          1   202279             3   
2         2        9327                  3          0   202279             3   
3         2       45918                  4          1   202279             3   
4         2       30035                  5          0   202279             3   

   orders_day_of_week  order_hour_of_day  days_since_prior_order _merge  
0                   5                  9                     8.0   both  
1                   5                  9                     8.0   both  
2                   5                  9                     8.0   both  
3                   5                  9                     8.0   both  
4                   5                  9                     8.0   both

In [89]:
# The '_merge' comlumn in the df_opcombined dataframe is not needed and might interfere with future data manipulations, so need to drop it
df_opcombined_1 = df_opcombined.drop(['_merge'], axis =1)

In [91]:
df_opcombined_1.head()

order_id  product_id  add_to_cart_order  reordered  user_id  order_number  \
0         2       33120                  1          1   202279             3   
1         2       28985                  2          1   202279             3   
2         2        9327                  3          0   202279             3   
3         2       45918                  4          1   202279             3   
4         2       30035                  5          0   202279             3   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                   5                  9                     8.0  
1                   5                  9                     8.0  
2                   5                  9                     8.0  
3                   5                  9                     8.0  
4                   5                  9                     8.0

# 03 Shape of the dataframes to be merged

In [94]:
# Checking the shape of the products data
df_prods.shape

(49670, 5)

In [96]:
# Checking the shape of the orders_products_combined data
df_opcombined_1.shape

(32434489, 9)

# 04 Merging products data to the combined orders and prior orders datafile

df_prods and df_opcombined datafames have product_id in common, so can be combined using this variable

For the data to be useful, products entered as product_id need to be identified, but also we are only interested in products that were ordered. So inner join would be most useful when merging. 

In [101]:
# Merging products and combined orders and prior orders detafiles
df_merged_large = df_opcombined_1.merge(df_prods, on = ['product_id'], indicator = True)

In [102]:
df_merged_large.head ()

order_id  product_id  add_to_cart_order  reordered  user_id  order_number  \
0         2       33120                  1          1   202279             3   
1         2       28985                  2          1   202279             3   
2         2        9327                  3          0   202279             3   
3         2       45918                  4          1   202279             3   
4         2       30035                  5          0   202279             3   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                   5                  9                     8.0   
1                   5                  9                     8.0   
2                   5                  9                     8.0   
3                   5                  9                     8.0   
4                   5                  9                     8.0   

            product_name  aisle_id  department_id  prices _merge  
0     Organic Egg Whites        86             16    11.3   both  
1  Michigan Organic Kale        83              4    13.4   both  
2          Garlic Powder       104             13     3.6   both  
3         Coconut Butter        19             13     8.4   both  
4      Natural Sweetener        17             13    13.7   both

In [105]:
#Checking the number of rows after the inner merge
df_merged_large['_merge'].value_counts()

_merge
both          32399732
left_only            0
right_only           0
Name: count, dtype: int64

32399732 rows were matched between the two dataframes

# 05  Exporting the merged datafile

In [109]:
# Exporting df_merged_large into prepared data folder as a pickle file
df_merged_large.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_final.pkl'))